### ISSUE TITLE
is it possible to specify different batch sizes for train and validation?

### ISSUE THREAD

**Issue by ophiry** [keras-team/keras#8550]
for train data, there are reasons to keep batches relatively small (batch size can effect training results)
however for the validation set, using a single reasonably big batch, allows simple use of metric functions that don't work well when averaging batches (precision \ recall \ f1)

**Comment by charlesbensimon** [comment]
If you don't specify any batch_size in your model definition (I think that you never have to, but I may be wrong), your model has a dynamic batch_size. You then use the batch_size parameter of the fit and predict methods of your keras models.

**Comment by ophiry** [comment]
I use fit_generator
the batch size of the train data is constant, though not explicitly defined in keras 
the validation data is sent as a numpy arrays (not as a generator) which are larger then the batch size of the train data

I see that event though no batch size was defined, keras will break the test data to batches in the size of the train set batch (I added a dummy metric that calculates the length of the batch to confirm this)

**Comment by charlesbensimon** [comment]
For that scenario, my advice is to do the epochs loop yourself :
for each epoch, you call fit_generator with epochs=1,
and then you call predict on your validation data with the desired batch size.

Another solution is to use a generator for the validation data, so you have control over the batch_size of your validation data

You could also use a custom Keras Callback
(in which you call predict inside on_epoch_end method)

**Comment by charlesbensimon** [comment]
But you're right, maybe a validation_batch_size parameter is missing for the case where validation_data is not a generator

**Comment by boulderZ** [comment]
I need this feature now. I am implementing a custom metric that calculates correlation. If I use the training batch size of 32 it is too small for my custom metric and I can get runs that are all zeros resulting in nan output for the validation metric. I would think this  could be a common issue for any metric that needs longer string of data to process accurately or in a stable way. Seems like this would be an easy feature to add?

**Comment by MatthewScholefield** [comment]
@boulderZ I'm not sure if this would help your issue, but when I ran into an `Nan` issue, I was able to fix it with `value / K.minimum(1.0, other_value)`. However, this didn't fix my issue of metric being averaged across the batch size. [This issue](https://github.com/keras-team/keras/issues/5794) has some relevant posts that details more of what @cbensimon and @ophiry are talking about.

**Comment by chmn106** [comment]
my advice is to define your validate function，use self define metric function
